pkg/ingester/recovery.go

```diff
@@ -0,0 +1,395 @@
+		for {
+			select {
+			case <-recoverer.Done():
+
+			case next, ok := <-input:
+				if !ok {
+					return
+				}
+				entries, ok := next.data.(RefEntries)
+				var err error
+				if !ok {
+					err = errors.Errorf("unexpected type (%T) when recovering WAL, expecting (%T)", next.data, entries)
+				}
+				if err == nil {
+					err = recoverer.Push(next.userID, entries)
+				}
+
+				// Pass the error back, but respect the quit signal.
+				if err != nil {
+					select {
+					case errCh <- err:
+					case <-recoverer.Done():
+					}
+					return
+				}
+			}
+		}
+	}
+
+	return recoverGeneric(
+		reader,
+		recoverer,
+		dispatch,
+		process,
+	)
+
+}
+
+func RecoverCheckpoint(reader WALReader, recoverer Recoverer) error {
+	dispatch := func(recoverer Recoverer, b []byte, inputs []chan recoveryInput, errCh <-chan error) error {
+		s := &Series{}
+		if err := decodeCheckpointRecord(b, s); err != nil {
+			return err
+		}
+
+		worker := int(s.Fingerprint % uint64(len(inputs)))
+		select {
+		case err := <-errCh:
+			return err
+
+		case inputs[worker] <- recoveryInput{
+			userID: s.UserID,
+			data:   s,
+		}:
+		}
+
+		return nil
+	}
+
+	process := func(recoverer Recoverer, input <-chan recoveryInput, errCh chan<- error) {
+		for {
+			select {
+			case <-recoverer.Done():
+
+			case next, ok := <-input:
+				if !ok {
+					return
+				}
+				series, ok := next.data.(*Series)
+				var err error
+				if !ok {
+					err = errors.Errorf("unexpected type (%T) when recovering WAL, expecting (%T)", next.data, series)
+				}
+				if err == nil {
+					err = recoverer.Series(series)
+				}
+
+				// Pass the error back, but respect the quit signal.
+				if err != nil {
+					select {
+					case errCh <- err:
+					case <-recoverer.Done():
+					}
+					return
+				}
+			}
+		}
+	}
+
+	return recoverGeneric(
+		reader,
+		recoverer,
+		dispatch,
+		process,
+	)
+}
+
+type recoveryInput struct {
+	userID string
+	data   interface{}
+}
+
+// recoverGeneric enables reusing the ability to recover from WALs of different types
+// by exposing the dispatch and process functions.
+// Note: it explicitly does not call the Recoverer.Close function as it's possible to layer
+// multiple recoveries on top of each other, as in the case of recovering from Checkpoints
+// then the WAL.
+func recoverGeneric(
+	reader WALReader,
+	recoverer Recoverer,
+	dispatch func(Recoverer, []byte, []chan recoveryInput, <-chan error) error,
+	process func(Recoverer, <-chan recoveryInput, chan<- error),
+) error {
+	var wg sync.WaitGroup
+	var lastErr error
+	nWorkers := recoverer.NumWorkers()
+
+	if nWorkers < 1 {
+		return errors.New("cannot recover with no workers")
+	}
+
+	errCh := make(chan error)
+	inputs := make([]chan recoveryInput, 0, nWorkers)
+	wg.Add(nWorkers)
+	for i := 0; i < nWorkers; i++ {
+		inputs = append(inputs, make(chan recoveryInput))
+
+		go func(input <-chan recoveryInput) {
+			process(recoverer, input, errCh)
+			wg.Done()
```

Comment:
good to have `defer wg.Done()` before `process`, just in case `process` panics or something?

Reply:
I like this idea a lot. In our case here it shouldn't make a difference because we're not catching a panic, but it makes the code more readable and feels safer to extend with your suggestion. Will add.